### ISSUE TITLE
Numerical instability in C51

### ISSUE THREAD

**Issue by vwxyzjn** [vwxyzjn/cleanrl#105]
C51 does a cross-entropy loss which could have numerical instability depending on the implementation. See [link](https://gist.github.com/zhanghuimeng/1fcd5aa6fdf162edce921248c7376d57) for an overview. Usually calculating the cross-entropy loss directly from the logits is more numerically stable. However, I am not sure how to do it exactly. 

Deepmind's dqn_zoo has an implementation seems to use the logits directly:

https://github.com/deepmind/rlax/blob/42bbcf97a69ef9b21cb88322b83169ade7930363/rlax/_src/value_learning.py#L703
https://github.com/deepmind/rlax/blob/42bbcf97a69ef9b21cb88322b83169ade7930363/rlax/_src/value_learning.py#L543

Personally, I am recording this issue but in practice often it's enough to do
```
loss = (-(target_pmfs * old_pmfs.clamp(min=1e-5, max=1-1e-5).log()).sum(-1)).mean()
# instead of 
# loss = (-(target_pmfs * old_pmfs.log()).sum(-1)).mean()
```

If anyone is interested in digging into this, that will be fantastic.

**Comment by ilovehhhyn** [comment]
Hi @vwxyzjn I could to take this if it’s still desired. I verified that current main still computes the C51 loss from clamped probabilities in both c51.py and c51_atari.py. I plan to expose the selected-action logits, compute the cross-entropy with log_softmax, and add CPU tests with extreme logits to verify finite losses/gradients and equivalence in normal ranges. Does that scope sound right?
